_plugins/link-checker.rb

```diff
@@ -1,219 +1,376 @@
-        @check_external_links = true
-        puts "LinkChecker: [Notice] The build will fail if any dead links are found"
-      else
-        puts "LinkChecker: [Notice] The build will fail if a dead internal link is found"
+  def self.verify(_site)
+    return unless @check_links
+
+    @base_url_matcher = %r{^#{@site.config["url"]}#{@site.baseurl}(/.*)$}.freeze
+    retry_hosts = {}
+
+    # Run atleast once
+    loop do
+      urls = @urls
+
+      # If its a retry
+      unless retry_hosts.empty?
+        # Get min sleep time in dict of timeouts and sleep
+        host_name, min_timeout_obj = @retry_timeouts_dict.min_by { |_k, v| v[:retry_timestamp] }
+        now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+        sleep_time = min_timeout_obj[:retry_timestamp] - now + @@retry_buffer
+
+        if sleep_time > 0
+          Jekyll.logger.info "LinkChecker: [Info] Going to sleep for #{sleep_time}".cyan
+          sleep(sleep_time)
+        end
+
+        # Get URLS to retry and clear from retry hash's
+        Jekyll.logger.info "LinkChecker: [Info] Retrying links for host #{host_name}".cyan
+        urls = retry_hosts[host_name].clone
+        @retry_timeouts_dict.delete(host_name)
+        retry_hosts.delete(host_name)
       end
+
+      # checl each url
+      # - valid URL: should not be failures but or in retry_hosts hash
+      # - invalid URL: should be failures but not in retry_hosts hash
+      # - retry URL: should not be failures, only in retry_hosts hash
+      urls.each do |url, pages|
+        Jekyll.logger.info "LinkChecker: [Info] Checking #{url}".cyan
+
+        valid_or_retry, metadata = check(url)
+        @failures << "Verify:: #{url}, linked to in ./#{pages.to_a.join(', ./')}" unless valid_or_retry
+
+        next unless @retry_external_links and metadata&.key?(:retry_host_name)
+
+        retry_host_name = metadata[:retry_host_name]
+        retry_hosts[retry_host_name] = [] unless retry_hosts.key?(retry_host_name)
+        retry_hosts[retry_host_name] << url
+      end
+
+      @retry_iteration += 1
+      break if !@retry_external_links or (@retry_iteration >= @@max_retry_iterations) or retry_hosts.empty?
     end
 
-    @urls.each do |url, pages|
-      @failures << "#{url}, linked to in ./#{pages.to_a.join(", ./")}" unless self.check(url)
+    unless @failures.empty?
+      msg = "Found #{@failures.size} dead link#{@failures.size > 1 ? 's' : ''}:\n#{@failures.join("\n")}"
     end
-    
-    msg = "Found #{@failures.size} dead link#{@failures.size > 1 ? 's' : ''}:\n#{@failures.join("\n")}" unless @failures.empty?
 
-    if @should_build_fatally
-      raise msg
+    unless retry_hosts.empty?
+      retry_msg = retry_hosts.map do |host, urls|
+        "Host:#{host}\n#{urls.map { |url| "- #{url}" }.join("\n")}\n"
+      end.join("\n")
+      msg = "Links we could not retry: \n#{retry_msg} \n#{msg}"
+    end
+
+    if !@failures.empty?
+      raise msg if @should_build_fatally
```

Comment:
I think a gentler way would be to just print an error message and exit(1). Writers who are running a link checker locally may attribute an exception to something not working in the script. If we're raising an exception, I believe it should be raise(msg) because msg is a string.

Reply:
Actually this doesn't work. If you `exit(1)` Jekyll will display this as an error, instead of displaying the actual error.

```
/home/runner/work/documentation-website/documentation-website/_plugins/link-checker.rb:195:in `exit': exit (SystemExit)
	from /home/runner/work/documentation-website/documentation-website/_plugins/link-checker.rb:195:in `verify'
	from /home/runner/work/documentation-website/documentation-website/_plugins/link-checker.rb:134:in `block in init'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/hooks.rb:103:in `block in trigger'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/hooks.rb:102:in `each'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/hooks.rb:102:in `trigger'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/site.rb:234:in `write'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/site.rb:82:in `process'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/command.rb:28:in `process_site'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:65:in `build'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:36:in `process'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `block in process_with_graceful_fai
```